File: js/src/value.ts
/*
 * Value — a decoded scalar or sub-blob MessagePack value.
 *
 * Mirrors msgpack::Value from the C++ Blob library. Integer values use bigint
 * so the full signed/unsigned 64-bit range round-trips exactly.
 */

import { utf8Decode, utf8Encode } from "./format.ts";

export const Type = {
  Nil: "null",
  True: "true",
  False: "false",
  Integer: "integer",
  Real: "real",
  Float32: "float32",
  String: "text",
  Binary: "binary",
  Array: "array",
  Map: "map",
  Ext: "ext",
  Timestamp: "timestamp",
} as const;
export type Type = (typeof Type)[keyof typeof Type];

export const IntWidth = {
  Auto: 0,
  Int8: 1,
  Int16: 2,
  Int32: 3,
  Int64: 4,
  Uint8: 5,
  Uint16: 6,
  Uint32: 7,
  Uint64: 8,
} as const;
export type IntWidth = (typeof IntWidth)[keyof typeof IntWidth];

export function typeStr(t: Type): string {
  return t;
}

const I64_MAX = (1n << 63n) - 1n;

export class Value {
  _type: Type;
  _int: bigint;
  _float: number;
  _str: Uint8Array;
  _blob: Uint8Array;
  _extType: number;
  _tsSec: bigint;
  _tsNsec: number;
  _intWidth: IntWidth;

  constructor() {
    this._type = Type.Nil;
    this._int = 0n;
    this._float = 0;
    this._str = new Uint8Array(0);
    this._blob = new Uint8Array(0);
    this._extType = 0;
    this._tsSec = 0n;
    this._tsNsec = 0;
    this._intWidth = IntWidth.Auto;
  }

  // ── accessors ─────────────────────────────────────────────────────
  type(): Type {
    return this._type;
  }
  isNil(): boolean {
    return this._type === Type.Nil;
  }
  asBool(): boolean {
    return this._type === Type.True;
  }
  asInt64(): bigint {
    if (this._type === Type.Integer) return BigInt.asIntN(64, this._int);
    if (this._type === Type.Real || this._type === Type.Float32) {
      return BigInt(Math.trunc(this._float));
    }
    if (this._type === Type.Timestamp) return this._tsSec;
    if (this._type === Type.True) return 1n;
    return 0n;
  }
  asUint64(): bigint {
    if (this._type === Type.Integer) return BigInt.asUintN(64, this._int);
    return 0n;
  }
  asDouble(): number {
    if (this._type === Type.Real || this._type === Type.Float32) return this._float;
    if (this._type === Type.Integer) return Number(this.asInt64());
    return 0;
  }
  asFloat(): number {
    if (this._type === Type.Float32) return this._float;
    if (this._type === Type.Real) return Math.fround(this._float);
    return 0;
  }
  asString(): string {
    if (this._type === Type.String) return utf8Decode(this._str);
    return "";
  }
  asBytes(): Uint8Array {
    return this._type === Type.String ? this._str : new Uint8Array(0);
  }
  blobData(): Uint8Array {
    return this._blob;
  }
  blobSize(): number {
    return this._blob.length;
  }
  extType(): number {
    return this._extType;
  }
  timestampSeconds(): bigint {
    return this._type === Type.Timestamp ? this._tsSec : 0n;
  }
  timestampNanoseconds(): number {
    return this._type === Type.Timestamp ? this._tsNsec : 0;
  }
  intWidth(): IntWidth {
    return this._intWidth;
  }

  // ── static constructors ───────────────────────────────────────────
  static nil(): Value {
    return new Value();
  }
  static boolean(b: boolean): Value {
    const v = new Value();
    v._type = b ? Type.True : Type.False;
    return v;
  }
  static integer(x: number | bigint): Value {
    const v = new Value();
    v._type = Type.Integer;
    v._int = BigInt(x);
    return v;
  }
  static unsignedInteger(x: number | bigint): Value {
    const v = new Value();
    v._type = Type.Integer;
    v._int = BigInt.asUintN(64, BigInt(x));
    if (v._int > I64_MAX) v._intWidth = IntWidth.Uint64;
    return v;
  }
  static real(d: number): Value {
    const v = new Value();
    v._type = Type.Real;
    v._float = d;
    return v;
  }
  static real32(f: number): Value {
    const v = new Value();
    v._type = Type.Float32;
    v._float = Math.fround(f);
    return v;
  }
  static string(s: string | Uint8Array): Value {
    const v = new Value();
    v._type = Type.String;
    v._str = typeof s === "string" ? utf8Encode(s) : s;
    return v;
  }
  static binary(data: Uint8Array): Value {
    const v = new Value();
    v._type = Type.Binary;
    v._blob = data;
    return v;
  }
  static ext(typeCode: number, data: Uint8Array): Value {
    const v = new Value();
    v._type = Type.Ext;
    v._extType = typeCode | 0;
    v._blob = data;
    return v;
  }
  static timestamp(seconds: number | bigint, nanoseconds: number = 0): Value {
    const v = new Value();
    v._type = Type.Timestamp;
    v._tsSec = BigInt(seconds);
    v._tsNsec = nanoseconds | 0;
    return v;
  }

  static _fixed(width: IntWidth, x: number | bigint): Value {
    const v = new Value();
    v._type = Type.Integer;
    v._int = BigInt(x);
    v._intWidth = width;
    return v;
  }
  static int8(x: number | bigint): Value {
    return Value._fixed(IntWidth.Int8, x);
  }
  static int16(x: number | bigint): Value {
    return Value._fixed(IntWidth.Int16, x);
  }
  static int32(x: number | bigint): Value {
    return Value._fixed(IntWidth.Int32, x);
  }
  static int64(x: number | bigint): Value {
    return Value._fixed(IntWidth.Int64, x);
  }
  static uint8(x: number | bigint): Value {
    return Value._fixed(IntWidth.Uint8, x);
  }
  static uint16(x: number | bigint): Value {
    return Value._fixed(IntWidth.Uint16, x);
  }
  static uint32(x: number | bigint): Value {
    return Value._fixed(IntWidth.Uint32, x);
  }
  static uint64(x: number | bigint): Value {
    return Value._fixed(IntWidth.Uint64, x);
  }
}

File: rust/src/value.rs
//! `Value` — a decoded scalar or sub-blob MessagePack value.
//!
//! Mirrors `msgpack::Value` from the C++ Blob library. Integer values are kept
//! as raw 64-bit bits so the full signed/unsigned range round-trips exactly.

/// Semantic type of a MessagePack element.
#[derive(Clone, Copy, PartialEq, Eq, Debug)]
pub enum Type {
    Nil,
    True,
    False,
    Integer,
    Real,
    Float32,
    String,
    Binary,
    Array,
    Map,
    Ext,
    Timestamp,
}

/// Integer encoding-width hint (forces a specific wire format).
#[derive(Clone, Copy, PartialEq, Eq, Debug)]
pub enum IntWidth {
    Auto,
    Int8,
    Int16,
    Int32,
    Int64,
    Uint8,
    Uint16,
    Uint32,
    Uint64,
}

/// Human-readable label for a [`Type`] (`"text"`, `"integer"`, …).
pub fn type_str(t: Type) -> &'static str {
    match t {
        Type::Nil => "null",
        Type::True => "true",
        Type::False => "false",
        Type::Integer => "integer",
        Type::Real => "real",
        Type::Float32 => "float32",
        Type::String => "text",
        Type::Binary => "binary",
        Type::Array => "array",
        Type::Map => "map",
        Type::Ext => "ext",
        Type::Timestamp => "timestamp",
    }
}

/// A decoded scalar or sub-blob value.
#[derive(Clone, Debug)]
pub struct Value {
    pub(crate) ty: Type,
    pub(crate) bits: u64, // integer bits (signed or unsigned) / timestamp seconds
    pub(crate) float: f64, // real / float32 payload
    pub(crate) bytes: Vec<u8>, // string / binary / ext payload (no header)
    pub(crate) ext_type: i8,
    pub(crate) ts_nsec: u32,
    pub(crate) int_width: IntWidth,
}

impl Default for Value {
    fn default() -> Self {
        Value {
            ty: Type::Nil,
            bits: 0,
            float: 0.0,
            bytes: Vec::new(),
            ext_type: 0,
            ts_nsec: 0,
            int_width: IntWidth::Auto,
        }
    }
}

impl Value {
    // ── accessors ─────────────────────────────────────────────────────
    pub fn get_type(&self) -> Type {
        self.ty
    }
    pub fn is_nil(&self) -> bool {
        self.ty == Type::Nil
    }
    pub fn as_bool(&self) -> bool {
        self.ty == Type::True
    }
    pub fn as_i64(&self) -> i64 {
        match self.ty {
            Type::Integer => self.bits as i64,
            Type::Real | Type::Float32 => self.float as i64,
            Type::Timestamp => self.bits as i64,
            Type::True => 1,
            _ => 0,
        }
    }
    pub fn as_u64(&self) -> u64 {
        if self.ty == Type::Integer {
            self.bits
        } else {
            0
        }
    }
    pub fn as_f64(&self) -> f64 {
        match self.ty {
            Type::Real | Type::Float32 => self.float,
            Type::Integer => self.as_i64() as f64,
            _ => 0.0,
        }
    }
    pub fn as_f32(&self) -> f32 {
        match self.ty {
            Type::Float32 => self.float as f32,
            Type::Real => self.float as f32,
            _ => 0.0,
        }
    }
    /// String payload bytes (verbatim; may not be valid UTF-8).
    pub fn as_bytes(&self) -> &[u8] {
        if self.ty == Type::String {
            &self.bytes
        } else {
            &[]
        }
    }
    /// String payload decoded lossily as UTF-8 (exact for valid UTF-8).
    pub fn as_string(&self) -> String {
        if self.ty == Type::String {
            String::from_utf8_lossy(&self.bytes).into_owned()
        } else {
            String::new()
        }
    }
    /// Binary / Ext payload (no header), or raw bytes for container values.
    pub fn blob_data(&self) -> &[u8] {
        &self.bytes
    }
    pub fn blob_size(&self) -> usize {
        self.bytes.len()
    }
    pub fn ext_type(&self) -> i8 {
        self.ext_type
    }
    pub fn timestamp_seconds(&self) -> i64 {
        if self.ty == Type::Timestamp {
            self.bits as i64
        } else {
            0
        }
    }
    pub fn timestamp_nanoseconds(&self) -> u32 {
        if self.ty == Type::Timestamp {
            self.ts_nsec
        } else {
            0
        }
    }
    pub fn int_width(&self) -> IntWidth {
        self.int_width
    }

    // ── static constructors ───────────────────────────────────────────
    pub fn nil() -> Value {
        Value::default()
    }
    pub fn boolean(b: bool) -> Value {
        Value {
            ty: if b { Type::True } else { Type::False },
            ..Default::default()
        }
    }
    pub fn integer(x: i64) -> Value {
        Value {
            ty: Type::Integer,
            bits: x as u64,
            ..Default::default()
        }
    }
    pub fn unsigned_integer(x: u64) -> Value {
        let mut v = Value {
            ty: Type::Integer,
            bits: x,
            ..Default::default()
        };
        if x > i64::MAX as u64 {
            v.int_width = IntWidth::Uint64;
        }
        v
    }
    pub fn real(d: f64) -> Value {
        Value {
            ty: Type::Real,
            float: d,
            ..Default::default()
        }
    }
    pub fn real32(f: f32) -> Value {
        Value {
            ty: Type::Float32,
            float: f as f64,
            ..Default::default()
        }
    }
    pub fn string(s: &str) -> Value {
        Value::string_bytes(s.as_bytes())
    }
    pub fn string_bytes(b: &[u8]) -> Value {
        Value {
            ty: Type::String,
            bytes: b.to_vec(),
            ..Default::default()
        }
    }
    pub fn binary(data: &[u8]) -> Value {
        Value {
            ty: Type::Binary,
            bytes: data.to_vec(),
            ..Default::default()
        }
    }
    pub fn ext(type_code: i8, data: &[u8]) -> Value {
        Value {
            ty: Type::Ext,
            ext_type: type_code,
            bytes: data.to_vec(),
            ..Default::default()
        }
    }
    pub fn timestamp(seconds: i64) -> Value {
        Value {
            ty: Type::Timestamp,
            bits: seconds as u64,
            ts_nsec: 0,
            ..Default::default()
        }
    }
    pub fn timestamp_ns(seconds: i64, nanoseconds: u32) -> Value {
        Value {
            ty: Type::Timestamp,
            bits: seconds as u64,
            ts_nsec: nanoseconds,
            ..Default::default()
        }
    }

    fn fixed(width: IntWidth, bits: u64) -> Value {
        Value {
            ty: Type::Integer,
            bits,
            int_width: width,
            ..Default::default()
        }
    }
    pub fn int8(x: i8) -> Value {
        Value::fixed(IntWidth::Int8, x as i64 as u64)
    }
    pub fn int16(x: i16) -> Value {
        Value::fixed(IntWidth::Int16, x as i64 as u64)
    }
    pub fn int32(x: i32) -> Value {
        Value::fixed(IntWidth::Int32, x as i64 as u64)
    }
    pub fn int64(x: i64) -> Value {
        Value::fixed(IntWidth::Int64, x as u64)
    }
    pub fn uint8(x: u8) -> Value {
        Value::fixed(IntWidth::Uint8, x as u64)
    }
    pub fn uint16(x: u16) -> Value {
        Value::fixed(IntWidth::Uint16, x as u64)
    }
    pub fn uint32(x: u32) -> Value {
        Value::fixed(IntWidth::Uint32, x as u64)
    }
    pub fn uint64(x: u64) -> Value {
        Value::fixed(IntWidth::Uint64, x)
    }
}
